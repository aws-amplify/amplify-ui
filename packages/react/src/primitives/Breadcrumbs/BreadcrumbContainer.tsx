import * as React from 'react';
import classNames from 'classnames';
import { ComponentClassName } from '@aws-amplify/ui';

import {
  Primitive,
  ForwardRefPrimitive,
  BreadcrumbsContainerProps,
  BaseBreadcrumbContainerProps,
} from '../types';
import { primitiveWithForwardRef } from '../utils/primitiveWithForwardRef';
import { View } from '../View';

const BreadcrumbContainerPrimitive: Primitive<
  BreadcrumbsContainerProps,
  'nav'
> = ({ className, children, ...rest }, ref) => {
  const componentClasses = classNames(
    ComponentClassName.Breadcrumbs,
    className
  );

  const ariaLabel = rest['aria-label'] ?? 'Breadcrumb';

  return (
    <View
      {...rest}
      as="nav"
      aria-label={ariaLabel}
      className={componentClasses}
      ref={ref}
    >
      <View as="ol" className={ComponentClassName.BreadcrumbsList}>
        {children}
      </View>
    </View>
  );
};

/**
 * [ðŸ“– Docs](https://ui.docs.amplify.aws/react/components/breadcrumbs)
 */
export const BreadcrumbContainer: ForwardRefPrimitive<
  BaseBreadcrumbContainerProps,
  'nav'
> = primitiveWithForwardRef(BreadcrumbContainerPrimitive);

BreadcrumbContainer.displayName = 'Breadcrumbs.Container';
