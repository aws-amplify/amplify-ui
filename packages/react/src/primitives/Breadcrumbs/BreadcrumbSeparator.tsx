import * as React from 'react';
import { classNames } from '@aws-amplify/ui';

import { ComponentClassName } from '@aws-amplify/ui';

import {
  BreadcrumbsSeparatorProps,
  ForwardRefPrimitive,
  Primitive,
} from '../types';
import { primitiveWithForwardRef } from '../utils/primitiveWithForwardRef';
import { View } from '../View';

const BreadcrumbSeparatorPrimitive: Primitive<
  BreadcrumbsSeparatorProps,
  'span'
> = ({ className, children = '/', as = 'span', ...rest }, ref) => {
  const ariaHidden = rest['aria-hidden'] ?? 'true';
  return (
    <View
      {...rest}
      as={as}
      ref={ref}
      aria-hidden={ariaHidden}
      className={classNames(ComponentClassName.BreadcrumbsSeparator, className)}
    >
      {children}
    </View>
  );
};

/**
 * [ðŸ“– Docs](https://ui.docs.amplify.aws/react/components/breadcrumbs)
 */
export const BreadcrumbSeparator: ForwardRefPrimitive<
  BreadcrumbsSeparatorProps,
  'span'
> = primitiveWithForwardRef(BreadcrumbSeparatorPrimitive);

BreadcrumbSeparator.displayName = 'Breadcrumbs.Separator';
