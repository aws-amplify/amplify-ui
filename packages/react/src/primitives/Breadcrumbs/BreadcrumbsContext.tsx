import * as React from 'react';

type BreadcrumbsContextType = {
  separator?: React.ReactNode;
  // We pass down isCurrent from
  // Breadcrumbs.Item to Breadcrumbs.Link
  // because we need that information on both elements
  isCurrent?: boolean;
};

export const BreadcrumbsContext = React.createContext<BreadcrumbsContextType>(
  {}
);

export const BreadcrumbsProvider = ({
  children,
  separator,
  isCurrent,
}: React.PropsWithChildren<BreadcrumbsContextType>): JSX.Element => {
  const value = React.useMemo(
    () => ({ separator, isCurrent }),
    [separator, isCurrent]
  );
  return (
    <BreadcrumbsContext.Provider value={value}>
      {children}
    </BreadcrumbsContext.Provider>
  );
};

export const useBreadcrumbsContext = (): BreadcrumbsContextType => {
  const context = React.useContext(BreadcrumbsContext);
  if (!context) {
    throw new Error(
      'useBreadcrumbsContext must be used within a BreadcrumbsProvider'
    );
  }
  return context;
};
