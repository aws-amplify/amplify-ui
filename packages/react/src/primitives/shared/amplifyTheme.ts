
/**
 * Defines a default Amplify SystemUI Theme
 * 
 * For all Chakra design tokens see:
 * https://github.com/chakra-ui/chakra-ui/tree/main/packages/theme/src/components
 */

// Button component ref: https://github.com/chakra-ui/chakra-ui/blob/main/packages/theme/src/components/button.ts
const Button = {
  baseStyle: {
    borderRadius: "var(--amplify-ui-border-radius-md)",
    color: "var(--amplify-ui-button-color-black-primary)",
    paddingInlineStart: "var(--amplify-ui-button-padding)",
    paddingInlineEnd: "var(--amplify-ui-button-padding)",
    fontWeight: "var(--amplify-ui-button-font-weight)",
    _disabled: {
      opacity: "var(--amplify-ui-button-disabled-opacity)",
      cursor: "var(--amplify-ui-button-disabled-cursor)"
    }
  },
  sizes: {
    lg: {
      height: "var(--amplify-ui-button-height-large)",
      fontSize: "var(--amplify-ui-button-font-size-large)"
    },
    md: {
      height: "var(--amplify-ui-button-height-medium)",
      fontSize: "var(--amplify-ui-button-font-size-medium)"
    },
    sm: {
      height: "var(--amplify-ui-button-height-small)",
      fontSize: "var(--amplify-ui-button-font-size-small)"
    }
  },
  variants: {
    primary: {
      bg: "var(--amplify-ui-button-color-bg-primary)",
      color: "var(--amplify-ui-button-color-text-primary)",
      _hover: {
        bg: "var(--amplify-ui-button-color-bg-primary-hover)",
        borderColor: "var(--amplify-ui-button-color-border-primary)",
        _disabled: {
          bg: "var(--amplify-ui-button-color-bg-primary)",
          color: "var(--amplify-ui-button-color-text-primary)",
        }
      },
    },
    secondary: {
      border: "1px solid var(--amplify-ui-button-color-border-secondary)",
      color: "var(--amplify-ui-button-color-text-secondary)",
      _hover: {
        borderColor: "var(--amplify-ui-button-color-border-secondary)",
        bg: "var(--amplify-ui-button-color-bg-secondary-hover)",
      },
    },
    tertiary: {
      bg: "var(--amplify-ui-button-color-bg-tertiary)",
      color: "var(--amplify-ui-button-color-text-tertiary)",
      _hover: {
        bg: "var(--amplify-ui-button-color-bg-tertiary-hover)",
      },
      _disabled: {
        bg: "var(--amplify-ui-button-color-bg-tertiary-hover)",
      }
    },
    link: {
      color: "var(--amplify-ui-button-color-text-link)",
      _hover: {
        textDecoration: "underline",
      }
    },
  },
  defaultProps: {
    variant: "secondary",
    size: "md"
  }
}

export const amplifyTheme = {
  components: {
    Button
  },
  config: {
    useSystemColorMode: false,
    initialColorMode: "light",
    cssVarPrefix: "amplify-theme"
  }
};
