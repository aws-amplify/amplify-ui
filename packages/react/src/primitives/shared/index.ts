export { ComponentClassNames, ComponentClassObject } from './constants';
