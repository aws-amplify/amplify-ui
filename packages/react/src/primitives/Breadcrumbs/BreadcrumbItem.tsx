import * as React from 'react';
import classNames from 'classnames';

import { ComponentClassNames } from '../shared/constants';
import {
  BaseBreadcrumbItemProps,
  BreadcrumbsItemProps,
  ForwardRefPrimitive,
  Primitive,
} from '../types';
import { View } from '../View';

const BreadcrumbItemPrimitive: Primitive<BreadcrumbsItemProps, 'li'> = (
  { className, children, as = 'li', ...rest },
  ref
) => {
  const componentClasses = classNames(
    ComponentClassNames.BreadcrumbsItem,
    className
  );

  return (
    <View {...rest} as={as} className={componentClasses} ref={ref}>
      {children}
    </View>
  );
};

/**
 * [ðŸ“– Docs](https://ui.docs.amplify.aws/react/components/breadcrumbs)
 */
export const BreadcrumbItem: ForwardRefPrimitive<
  BaseBreadcrumbItemProps,
  'span'
> = React.forwardRef(BreadcrumbItemPrimitive);

BreadcrumbItem.displayName = 'Breadcrumbs.Item';
