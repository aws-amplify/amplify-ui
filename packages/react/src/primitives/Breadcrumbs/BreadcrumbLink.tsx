import * as React from 'react';
import classNames from 'classnames';

import { classNameModifierByFlag } from '../shared/utils';
import {
  BreadcrumbsLinkProps,
  ForwardRefPrimitive,
  Primitive,
  BaseBreadcrumbLinkProps,
} from '../types';
import { ComponentClassNames } from '../shared/constants';
import { Link } from '../Link';
import { Text } from '../Text';

const BreadcrumbLinkPrimitive: Primitive<BreadcrumbsLinkProps, 'a'> = (
  { className, children, href, isCurrent, ...rest },
  ref
) => {
  const componentClasses = classNames(
    ComponentClassNames.BreadcrumbsLink,
    classNameModifierByFlag(
      ComponentClassNames.BreadcrumbsLink,
      'current',
      isCurrent
    ),
    className
  );

  if (isCurrent) {
    const ariaCurrent = rest['aria-current'] ?? 'page';
    const as = rest.as ?? 'span';

    return (
      <Text
        {...rest}
        as={as}
        aria-current={ariaCurrent}
        className={componentClasses}
        ref={ref}
      >
        {children}
      </Text>
    );
  } else {
    return (
      <Link {...rest} className={componentClasses} ref={ref} href={href}>
        {children}
      </Link>
    );
  }
};

/**
 * [ðŸ“– Docs](https://ui.docs.amplify.aws/react/components/breadcrumbs)
 */
export const BreadcrumbLink: ForwardRefPrimitive<BaseBreadcrumbLinkProps, 'a'> =
  React.forwardRef(BreadcrumbLinkPrimitive);

BreadcrumbLink.displayName = 'Breadcrumbs.Link';
