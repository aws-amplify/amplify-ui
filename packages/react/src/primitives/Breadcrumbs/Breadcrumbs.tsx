import * as React from 'react';

import {
  BaseBreadcrumbProps,
  BreadcrumbsProps,
  Primitive,
  ForwardRefPrimitive,
} from '../types';

import { BreadcrumbItem } from './BreadcrumbItem';
import { BreadcrumbLink } from './BreadcrumbLink';
import { BreadcrumbSeparator } from './BreadcrumbSeparator';
import { BreadcrumbContainer } from './BreadcrumbContainer';

const BreadcrumbsPrimitive: Primitive<BreadcrumbsProps, 'nav'> = (
  { className, items, separator = <BreadcrumbSeparator />, ...rest },
  ref
) => {
  const ariaLabel = rest['aria-label'] ?? 'Breadcrumb';

  return (
    <BreadcrumbContainer
      {...rest}
      aria-label={ariaLabel}
      className={className}
      ref={ref}
    >
      {items?.map(({ href, label }, idx) => {
        const isCurrent = items.length - 1 === idx;
        return (
          <BreadcrumbItem key={href}>
            <BreadcrumbLink href={href} isCurrent={isCurrent}>
              {label}
            </BreadcrumbLink>
            {/* Don't show separator if item isCurrent */}
            {isCurrent ? null : separator}
          </BreadcrumbItem>
        );
      })}
    </BreadcrumbContainer>
  );
};

type BreadcrumbsType = ForwardRefPrimitive<BaseBreadcrumbProps, 'nav'> & {
  Link: typeof BreadcrumbLink;
  Item: typeof BreadcrumbItem;
  Separator: typeof BreadcrumbSeparator;
  Container: typeof BreadcrumbContainer;
};

/**
 * [ðŸ“– Docs](https://ui.docs.amplify.aws/react/components/breadcrumbs)
 */
const Breadcrumbs: BreadcrumbsType = Object.assign(
  React.forwardRef(BreadcrumbsPrimitive),
  {
    Item: BreadcrumbItem,
    Link: BreadcrumbLink,
    Separator: BreadcrumbSeparator,
    Container: BreadcrumbContainer,
  }
);

Breadcrumbs.displayName = 'Breadcrumbs';

export { Breadcrumbs };
