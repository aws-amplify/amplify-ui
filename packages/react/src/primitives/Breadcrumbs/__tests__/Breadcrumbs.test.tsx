import * as React from 'react';
import { render } from '@testing-library/react';
import { Breadcrumbs } from '../Breadcrumbs';
import { ComponentClassNames } from '../../shared';

const breadcrumbs = [
  {
    href: 'test',
    label: 'test',
  },
  {
    label: 'test',
  },
];

describe('Breadcrumbs:', () => {
  it('should match snapshot', () => {
    const { container } = render(<Breadcrumbs items={breadcrumbs} />);
    expect(container).toMatchSnapshot();
  });

  it('renders the proper accessibility attributes', () => {
    const { container } = render(<Breadcrumbs items={breadcrumbs} />);
    // https://www.w3.org/WAI/ARIA/apg/patterns/breadcrumb/examples/breadcrumb/
    expect(
      container.querySelector(`.${ComponentClassNames.Breadcrumbs}`)
    ).toHaveAttribute('aria-label', 'Breadcrumb');
  });

  it('can render custom separators', () => {
    const { container } = render(
      <Breadcrumbs
        items={breadcrumbs}
        separator={<span className="test">|</span>}
      />
    );

    expect(container.querySelector('.test')).toBeInTheDocument();
  });

  it('can apply custom classNames', () => {
    const { container } = render(
      <Breadcrumbs.Container className="custom-breadcrumbs">
        <Breadcrumbs.Item className="custom-breadcrumbs__item">
          <Breadcrumbs.Link className="custom-breadcrumbs__link" href="/test">
            Test
          </Breadcrumbs.Link>
        </Breadcrumbs.Item>
      </Breadcrumbs.Container>
    );
    expect(container.querySelector('.custom-breadcrumbs')).toBeInTheDocument();
    expect(
      container.querySelector('.custom-breadcrumbs__item')
    ).toBeInTheDocument();
    expect(
      container.querySelector('.custom-breadcrumbs__link')
    ).toBeInTheDocument();
  });

  it('should apply proper aria attributes to current link', () => {
    const { container } = render(<Breadcrumbs items={[{ label: 'Test' }]} />);
    const linkElement = container.querySelector(
      `.${ComponentClassNames.BreadcrumbsLink}`
    );
    expect(linkElement).toHaveAttribute('aria-current', 'page');
  });
});
