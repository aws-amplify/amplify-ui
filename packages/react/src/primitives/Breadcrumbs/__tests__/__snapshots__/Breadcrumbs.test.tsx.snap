// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs: should match snapshot 1`] = `
<div>
  <nav
    aria-label="Breadcrumb"
    class="amplify-breadcrumbs"
  >
    <ol
      class="amplify-breadcrumbs__list"
    >
      <li
        class="amplify-breadcrumbs__item"
      >
        <a
          class="amplify-link amplify-breadcrumbs__link"
          href="test"
        >
          test
        </a>
        <span
          aria-hidden="true"
          class="amplify-breadcrumbs__separator"
        >
          /
        </span>
      </li>
      <li
        class="amplify-breadcrumbs__item"
      >
        <span
          aria-current="page"
          class="amplify-text amplify-breadcrumbs__link amplify-breadcrumbs__link--current"
        >
          test
        </span>
      </li>
    </ol>
  </nav>
</div>
`;
